Mathlib/Algebra/Group/Hom/CompTypeclasses.lean

```diff
@@ -0,0 +1,105 @@
+Authors: Antoine Chambert-Loir
+-/
+
+
+import Mathlib.Logic.Function.CompTypeclasses
+import Mathlib.Algebra.Group.Hom.Defs
+
+/-!
+# Propositional typeclasses on several monoid homs
+
+This file contains typeclasses used in the definition of equivariant maps,
+in the spirit what was initially developed by Frédéric Dupuis and Heather Macbeth for linear maps.
+However, we do not expect that all maps should be guessed automatically,
+as it happens for linear maps.
+
+If `φ`, `ψ`… are monoid homs and `M`, `N`… are monoids, we add two instances:
+* `MonoidHom.CompTriple φ ψ χ`, which expresses that `ψ.comp φ = χ`
+* `MonoidHom.IsId φ`, which expresses that `φ = id`
+
+Some basic lemmas are proved:
+* `MonoidHom.CompTriple.comp` asserts `MonoidHom.CompTriple φ ψ (ψ.comp φ)`
+* `MonoidHom.CompTriple.id_comp` asserts `MonoidHom.CompTriple φ ψ ψ`
+  in the presence of `MonoidHom.IsId φ`
+* its variant `MonoidHom.CompTriple.comp_id`
+
+TODO :
+* align with RingHomCompTriple
+* probably rename MonoidHom.CompTriple as MonoidHomCompTriple
+(or, on the opposite, rename RingHomCompTriple as RingHom.CompTriple)
+* does one need AddHom.CompTriple ?
+
+-/
+
+section MonoidHomCompTriple
+
+namespace MonoidHom
+
+/-- Class of composing triples -/
+class CompTriple  {M N P : Type*} [Monoid M] [Monoid N] [Monoid P]
+  (φ : M →* N) (ψ : N →* P) (χ : outParam (M →* P)) : Prop where
+  /-- The maps form a commuting triangle -/
+  comp_eq : ψ.comp φ = χ
+
+attribute [simp] CompTriple.comp_eq
+
+namespace CompTriple
+
+variable {M' : Type*} [Monoid M']
+variable {M N P : Type*} [Monoid M] [Monoid N] [Monoid P]
+
+/-- Class of Id maps -/
+class IsId (σ : M →* M) : Prop where
+  eq_id : σ = MonoidHom.id M
+
+instance instIsId {M : Type*} [Monoid M] : IsId (MonoidHom.id M) where
+  eq_id := rfl
+
+instance {σ : M →* M} [h : _root_.CompTriple.IsId σ] : IsId σ  where
+  eq_id := by ext; exact _root_.congr_fun h.eq_id _
+
+instance instComp_id {N P : Type*} [Monoid N] [Monoid P]
+    {φ : N →* N} [IsId φ] {ψ : N →* P} :
+    CompTriple φ ψ ψ where
+  comp_eq := by simp only [IsId.eq_id, MonoidHom.comp_id]
+
+instance instId_comp {M N : Type*} [Monoid M] [Monoid N]
+    {φ : M →* N} {ψ : N →* N} [IsId ψ] :
+    CompTriple φ ψ φ where
+  comp_eq := by simp only [IsId.eq_id, MonoidHom.id_comp]
+
+lemma comp_inv {φ : M →* N} {ψ : N →* M} (h : Function.RightInverse φ ψ)
+    {χ : M →* M} [IsId χ] :
+    CompTriple φ ψ χ where
+  comp_eq := by
+    simp only [IsId.eq_id, ← DFunLike.coe_fn_eq, coe_comp, h.id]
+    rfl
+
+instance instRootCompTriple {φ : M →* N} {ψ : N  →* P} {χ : M →* P} [κ : CompTriple φ ψ χ] :
+    _root_.CompTriple φ ψ χ where
+  comp_eq := by rw [← MonoidHom.coe_comp, κ.comp_eq]
+
+instance (priority := 10) instComp {φ : M →* N} {ψ : N →* P} :
+    CompTriple φ ψ (ψ.comp φ) where
+  comp_eq := rfl
+
+lemma comp_apply
+    {φ : M →* N} {ψ : N →* P} {χ : M →* P} (h : CompTriple φ ψ χ) (x : M) :
+    ψ (φ x) = χ x := by
```

Comment:
```suggestion
lemma comp_apply
    {φ : M →* N} {ψ : N →* P} {χ : M →* P} [CompTriple φ ψ χ] (x : M) :
    ψ (φ x) = χ x := by
```
Is there a reason why we don't do this here? It feels like having the instance found by typeclass inference is the whole point, no? Or perhaps we want both versions...? (It's a bit hard for me to tell, since I don't really know what the actual applications look like.)

Reply:
The actual applications are exactly the same as for semilinear maps, but for maps of types endowed with `Mul` or `MulAction`.